test/test_optim.py

```diff
@@ -67,7 +67,8 @@ def eval(params, sparse_grad, w):
                 v = torch.DoubleTensor([y - y / 4., y / 4.])
             x = sparse.DoubleTensor(i, v, torch.Size([2]))
             if sparse_grad:
-                params.grad.data = x
+                with torch.no_grad():
+                    params.grad = x
```

Comment:
In the old `variable.set_data(data_tensor)` method, it was ok to pass a sparse `data_tensor` to a dense `variable` (or vice versa). In the new `variable.set_data(data_tensor)` method, this behavior is not possible to implement (we can't just swap out the variable's TensorImpl and put in a SparseTensorImpl), we instead choose to disallow this behavior, and fix places where this behavior is expected (which only occurs here and in `accumulate_grad.cpp`).